File: golang/roman_to_int.go
package main

/*
Space	: O(1)
Time	: O(n)
*/

func romanToInt(s string) int {
	rmap := map[byte]int{
		'I': 1,
		'V': 5,
		'X': 10,
		'L': 50,
		'C': 100,
		'D': 500,
		'M': 1000,
	}
	ans := 0
	n := len(s)

	for i := 0; i < n-1; i++ {
		if rmap[s[i]] >= rmap[s[i+1]] {
			ans += rmap[s[i]]
		} else {
			ans -= rmap[s[i]]
		}
	}

	ans += rmap[s[n-1]]
	return ans
}

File: python3/roman_to_int.py
"""
Space : O(7)
Time  : O(n)
"""


class Solution:
    def romanToInt(self, s: str) -> int:
        romanmap = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000,
        }
        n = len(s)
        ans = 0

        for idx in range(n-1):
            if romanmap[s[idx]] >= romanmap[s[idx+1]]:
                ans += romanmap[s[idx]]
            else:
                ans -= romanmap[s[idx]]

        ans += romanmap[s[-1]]
        return ans
